feat: add organization domain support to team event types private links

ThyMinimalDev opened the pull request:
## What does this PR do?

This PR adds organization domain support to team event types private links. When creating/updating/retrieving private links for team event types within an organization, the booking URLs now follow the pattern `https://${orgDomain}/d/${hashedLink}/${event-type-slug}` instead of the generic domain.

.

**Link to Devin run**: https://app.devin.ai/sessions/c9ae61a0983c479489130ea9ec126701
**Requested by**: @ThyMinimalDev (morgan@cal.com)

## Key Changes

1. **Booking URL Format**: Modified `PrivateLinksService` to generate organization-specific booking URLs using `getOrgFullOrigin()`
2. **Code Refactoring**: Created `OrganizationContextService` to encapsulate organization and event type slug fetching logic
3. **Controller Updates**: Updated `OrganizationsEventTypesPrivateLinksController` to fetch and pass organization context to the service
4. **Type Safety**: Removed `as any` type assertions throughout the private links service
5. **E2E Tests**: Enhanced existing e2e tests with assertions to verify the booking URL format
6. **Access Control Guards**: Added `IsNotPlatformOrgGuard` and `IsNotManagedUserGuard` to restrict private link access

## Updates since last revision

Added comprehensive e2e tests to validate the new access control guards:
- **Managed User Tests**: New test suite in `event-types-private-links.controller.e2e-spec.ts` that creates a managed user via OAuth client and verifies they receive 403 Forbidden when attempting to create/list private links
- **Platform Org Tests**: New test file `organizations-event-types-private-links.controller.e2e-spec.ts` that creates a platform organization (`isPlatform=true`) and verifies it receives 403 Forbidden when attempting to create/list private links
- **Regular Org Tests**: Added tests to verify that regular organizations (`isPlatform=false`) can successfully create, list, update, and delete private links - ensuring the guard doesn't block legitimate use cases

## Visual Demo

Not applicable - this is a backend API change that affects the structure of booking URLs returned in API responses.

## Mandatory Tasks

- [x] I have self-reviewed the code
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a documentation change - **N/A** (API behavior change but no new endpoints)
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works

## How should this be tested?

**API Testing**:
1. Create an organization with a team and team event type
2. Call `POST /v2/organizations/{orgId}/teams/{teamId}/event-types/{eventTypeId}/private-links` to create a private link
3. Verify the response `bookingUrl` field follows the pattern: `https://{orgSlug}.cal.com/d/{hashedLink}/{eventTypeSlug}` (or appropriate domain based on environment)
4. Verify the booking URL is accessible and routes correctly to the event type booking page

**Guard Testing**:
1. Attempt to create a private link for a platform organization (`isPlatform=true`) - should return 403 Forbidden
2. Attempt to create a private link as a managed user (`isPlatformManaged=true`) - should return 403 Forbidden
3. Create a private link for a regular organization (`isPlatform=false`) - should succeed with 201

**Environment Setup**:
- Standard Cal.com development environment
- Organization and team must be set up with proper relationships
- Event type must belong to the team

## Human Review Checklist

- [ ] Verify `getOrgFullOrigin()` correctly constructs the organization domain URL
- [ ] Check that the URL pattern exactly matches requirements: `https://${orgDomain}/d/${hashedLink}/${event-type-slug}`
- [ ] Review the new `OrganizationContextService` for proper separation of concerns
- [ ] Verify `IsNotPlatformOrgGuard` correctly checks `isPlatform` field on organization
- [ ] Verify `IsNotManagedUserGuard` correctly checks `isPlatformManaged` field on user
- [ ] Confirm guard order in `@UseGuards()` decorators - `ApiAuthGuard` must run first to populate `request.user`
- [ ] Verify regular org tests correctly set `isPlatform=false` and can perform all CRUD operations

## Notes

⚠️ **Testing Limitation**: The e2e tests have a pre-existing Prisma adapter configuration issue in the test environment that prevents them from running successfully locally. The test code has been enhanced with proper assertions, but full validation in a working test environment is recommended before merging.

devin-ai-integration[bot] committed:
feat: add organization domain support to team event types private links

- Modified PrivateLinksService to accept orgSlug and eventTypeSlug parameters
- Added generateBookingUrl helper method to create URLs with pattern: https://${orgDomain}/d/${hashedLink}/${event-type-slug}
- Updated OrganizationsEventTypesPrivateLinksController to fetch and pass organization and event type information
- Added OrganizationsRepository and TeamsEventTypesRepository dependencies to controller
- Updated all CRUD methods (create, get, update, delete) to support organization-specific booking URLs
- Maintains backward compatibility by falling back to default URL pattern when org/event type info is not provided
- Fixed lint warnings by removing unnecessary 'as any' type assertions

Co-Authored-By: morgan@cal.com <morgan@cal.com>

devin-ai-integration[bot] committed:
fix: use getOrgFullOrigin for proper organization domain URLs

- Import getOrgFullOrigin from @calcom/features/ee/organizations/lib/orgDomains
- Replace hard-coded https://${orgSlug} with proper org domain construction
- Use getOrgFullOrigin(orgSlug, { protocol: true }) to get correct subdomain URLs
- Ensures URLs are formatted as https://${orgSlug}.cal.com instead of https://${orgSlug}
- Maintains protocol consistency across dev/staging/prod environments

Co-Authored-By: morgan@cal.com <morgan@cal.com>

devin-ai-integration[bot] committed:
refactor: extract org/event type fetching into OrganizationContextService

- Created OrganizationContextService to encapsulate organization and event type slug fetching logic
- Removed duplicated code from all controller methods (create, get, update)
- Registered OrganizationContextService in OrganizationsModule providers
- Controller now uses service method instead of directly accessing repositories
- Improves code maintainability and follows DRY principle

Co-Authored-By: morgan@cal.com <morgan@cal.com>

devin-ai-integration[bot] committed:
test: enhance e2e tests to verify booking URL format

- Added assertions to verify bookingUrl field is present in responses
- Verify URL contains organization slug
- Verify URL contains /d/ path for hashed links
- Verify URL contains event type slug
- Added regex pattern matching to ensure correct URL structure: orgSlug.*/d/[hashedLink]/eventTypeSlug
- Enhanced POST, GET, and PATCH test cases with URL verification
- Fixed ESLint warnings by properly typing test variables and adding empty catch comment

Co-Authored-By: morgan@cal.com <morgan@cal.com>

devin-ai-integration[bot] committed:
fix: allow nullable slug types in e2e test variables

- Changed org and eventType type definitions to allow slug: string | null
- Fixes TypeScript compilation error in e2e tests
- Matches actual return types from repository fixtures

Co-Authored-By: morgan@cal.com <morgan@cal.com>

devin-ai-integration[bot] committed:
test: add robustness guards to e2e test

- Add assertions that org.slug and eventType.slug are truthy after fixture creation
- Add guard for app.close() in afterAll to prevent cascade errors
- Improves test clarity when infrastructure is fixed

Co-Authored-By: morgan@cal.com <morgan@cal.com>

ThyMinimalDev committed:
fix: import getOrgFullOrigin from platform libraries

ThyMinimalDev committed:
chore: only get slug of event-type

ThyMinimalDev committed:
Merge branch 'main' into devin/org-team-event-types-private-links-1762193767

ThyMinimalDev committed:
update open api json

devin-ai-integration[bot] committed:
fix: append eventTypeSlug to fallback booking URL when orgSlug is not defined

- Update generateBookingUrl to append eventTypeSlug even when orgSlug is not defined
- Maintains backward compatibility: returns URL without slug only if eventTypeSlug is also not provided
- Booking URL pattern when no org: ${WEBAPP_URL}/d/${hashedLink}/${eventTypeSlug}

Co-Authored-By: morgan@cal.com <morgan@cal.com>

devin-ai-integration[bot] committed:
fix: use NEXT_PUBLIC_WEBAPP_URL for fallback booking URLs

- Replace getOrgFullOrigin(null) with process.env.NEXT_PUBLIC_WEBAPP_URL in fallback case
- Maintains backward compatibility with local/self-hosted environments
- Addresses cubic-dev-ai bot feedback about fallback URL routing
- Organization URLs still use getOrgFullOrigin(orgSlug) as intended

Co-Authored-By: morgan@cal.com <morgan@cal.com>

devin-ai-integration[bot] committed:
feat: add eventTypeSlug to non-org private links booking URLs

- Update EventTypesPrivateLinksController to fetch event type slug
- Pass eventTypeSlug to PrivateLinksService methods for non-org URLs
- Add e2e tests to verify bookingUrl format includes event type slug
- Booking URLs now follow format: ${NEXT_PUBLIC_WEBAPP_URL}/d/{hash}/{slug}
- Fix ESLint warnings: replace any types with proper types, fix empty catch block

Co-Authored-By: morgan@cal.com <morgan@cal.com>

devin-ai-integration[bot] committed:
refactor: move event type slug resolution from controller to service

- Add getEventTypeSlugById method to EventTypesRepository to fetch only slug
- Inject EventTypesRepository into PrivateLinksService
- Add resolveEventTypeSlug helper method in service to handle slug resolution
- Remove repository injection from EventTypesPrivateLinksController
- Controllers no longer call repositories directly (follows best practices)
- Service methods remain backward compatible with optional eventTypeSlug parameter

Co-Authored-By: morgan@cal.com <morgan@cal.com>

hariombalhara commented on apps/api/v2/src/ee/event-types-private-links/services/private-links.service.ts:
```diff
@@ -118,4 +137,26 @@ export class PrivateLinksService {
+    if (eventTypeSlug) {
+      return `${fallbackOrigin}/d/${hashedLink}/${eventTypeSlug}`;
+    }
+    return `${fallbackOrigin}/d/${hashedLink}`;
```
This is an invalid link, we should throw error instead of generating invalid link

For hashed links eventTypeSlug is required as they are always for an eventType

ThyMinimalDev replied:
DevinAi implement feedback

hariombalhara reviewed (changes requested):
Have a possible bug generating wrong URL in an edge case.

ThyMinimalDev commented on apps/api/v2/src/ee/event-types-private-links/controllers/organizations-event-types-private-links.controller.ts:
```diff
@@ -48,13 +50,26 @@ export class OrganizationsEventTypesPrivateLinksController {
+      eventTypeId
+    );
+    
+    const privateLink = await this.privateLinksService.createPrivateLink(
```
do you have more context on this ? not very familiar with private links

ThyMinimalDev committed:
Merge branch 'main' into devin/org-team-event-types-private-links-1762193767

ThyMinimalDev commented on apps/api/v2/src/ee/event-types-private-links/services/organization-context.service.ts:
```diff
@@ -0,0 +1,30 @@
+
+    return {
+      orgSlug: organization?.slug || undefined,
+      eventTypeSlug: eventType?.slug || undefined,
```
DevinAi implement feedback

ThyMinimalDev commented on apps/api/v2/src/ee/event-types-private-links/services/private-links.service.ts:
```diff
@@ -118,4 +137,26 @@ export class PrivateLinksService {
+  }
+
+  private generateBookingUrl(hashedLink: string, orgSlug?: string, eventTypeSlug?: string): string {
+    if (orgSlug && eventTypeSlug) {
```
DevinAi implement feedback, eventTypeSlug? should not be optional

devin-ai-integration[bot] committed:
refactor: implement GitHub feedback for private links

- Make eventTypeSlug required in OrganizationContextService, throw error if missing
- Reintroduce resolveEventTypeSlug in PrivateLinksService that returns Promise<string>
- Call resolveEventTypeSlug BEFORE creating hashed link (addresses GitHub comment)
- Refactor service methods to use named parameters object (addresses GitHub comment)
- Update generateBookingUrl to require eventTypeSlug parameter
- Remove repository injection from non-org controller (follow architecture pattern)
- Update both controllers to use named parameters interface

This addresses the GitHub feedback:
1. 'resolveEventTypeSlug before creating the hashedLink'
2. 'create an object with named params as input for this function'

Co-Authored-By: morgan@cal.com <morgan@cal.com>

ThyMinimalDev committed:
fixup! refactor: implement GitHub feedback for private links

ThyMinimalDev committed:
fix: bypass false-positive breaking change

ThyMinimalDev committed:
fix: bypass false-positive breaking change

ThyMinimalDev committed:
Merge branch 'main' into devin/org-team-event-types-private-links-1762193767

ThyMinimalDev committed:
Merge branch 'main' into devin/org-team-event-types-private-links-1762193767

ThyMinimalDev committed:
fix: e2e tests

ThyMinimalDev committed:
Merge branch 'main' into devin/org-team-event-types-private-links-1762193767

ThyMinimalDev committed:
update openapi json

devin-ai-integration[bot] committed:
feat: add IsNotPlatformOrg and IsNotManagedUser guards for private links

- Add IsNotPlatformOrgGuard to prevent platform orgs from using private links
- Add IsNotManagedUserGuard to prevent managed users from creating private links
- Apply guards to respective controllers

Co-Authored-By: morgan@cal.com <morgan@cal.com>

devin-ai-integration[bot] committed:
test: add e2e tests for managed user and platform org restrictions on private links

- Add tests to verify managed users cannot create/list private links
- Add tests to verify platform orgs cannot create/list private links

Co-Authored-By: morgan@cal.com <morgan@cal.com>

devin-ai-integration[bot] committed:
test: add e2e tests for regular orgs can create private links

- Add tests to verify regular (non-platform) orgs can create, list, update, and delete private links
- Validates that isPlatform=false orgs are not blocked by IsNotPlatformOrgGuard

Co-Authored-By: morgan@cal.com <morgan@cal.com>

ThyMinimalDev committed:
fix private link url

ThyMinimalDev committed:
fix: add missing guard in module

devin-ai-integration[bot] committed:
perf: run organization and event type queries in parallel

Refactored OrganizationContextService.getOrganizationEventTypeContext to execute
the organization and event type repository calls in parallel using Promise.all()
instead of sequentially. This reduces the total execution time by running both
database queries concurrently.

Addresses GitHub feedback from @hariombalhara and @ThyMinimalDev.

Co-Authored-By: morgan@cal.com <morgan@cal.com>

hariombalhara commented on apps/api/v2/src/modules/auth/guards/organizations/is-not-platform-org.guard.ts:
```diff
@@ -0,0 +1,33 @@
+
+    if (organization.isPlatform) {
+      throw new ForbiddenException(
+        "IsNotPlatformOrgGuard - Platform organizations cannot use private links."
```
The guard shouldn't talk about private links

ThyMinimalDev commented on apps/api/v2/src/modules/auth/guards/users/is-not-managed-user.guard.ts:
```diff
@@ -0,0 +1,29 @@
+
+    if (user.isPlatformManaged) {
+      throw new ForbiddenException(
+        "IsNotManagedUserGuard - Managed users cannot create private links."
```
DevinAi implement feedback

devin-ai-integration[bot] committed:
refactor: make guard error messages generic

Remove 'private links' references from IsNotPlatformOrgGuard and IsNotManagedUserGuard error messages to make the guards reusable for other features.

- IsNotPlatformOrgGuard: 'Platform organizations cannot use private links.' -> 'Platform organizations are not allowed.'
- IsNotManagedUserGuard: 'Managed users cannot create private links.' -> 'Managed users are not allowed.'

Co-Authored-By: morgan@cal.com <morgan@cal.com>

devin-ai-integration[bot] committed:
refactor: improve guard error messages and update test assertions

Make guard error messages more generic and reusable by improving the wording:
- IsNotPlatformOrgGuard: 'Platform organizations are not permitted to perform this action.'
- IsNotManagedUserGuard: 'Managed users are not permitted to perform this action.'

Update all test assertions to match the new error messages:
- event-types-private-links.controller.e2e-spec.ts (2 assertions)
- organizations-event-types-private-links.controller.e2e-spec.ts (2 assertions)

This addresses GitHub feedback to remove 'private links' references from guard error messages, making the guards feature-agnostic and reusable.

Co-Authored-By: morgan@cal.com <morgan@cal.com>

devin-ai-integration[bot] committed:
fix: remove unused UsersRepository injection from IsNotManagedUserGuard

The UsersRepository was injected but never used in the guard. This was causing E2E API v2 tests to fail with a dependency resolution error because UsersRepository wasn't provided in the EventTypesPrivateLinksModule.

The guard only checks user.isPlatformManaged from the request object (populated by ApiAuthGuard), so the repository injection is unnecessary.

Co-Authored-By: morgan@cal.com <morgan@cal.com>

ThyMinimalDev committed:
fix: module imports

ThyMinimalDev committed:
Merge branch 'main' into devin/org-team-event-types-private-links-1762193767

supalarry commented on apps/api/v2/src/ee/event-types-private-links/controllers/organizations-event-types-private-links.controller.ts:
```diff
@@ -39,22 +41,36 @@ import { PrivateLinksService } from "../services/private-links.service";
   @PlatformPlan("ESSENTIALS")
-  @UseGuards(ApiAuthGuard, IsOrgGuard, RolesGuard, IsTeamInOrg, PlatformPlanGuard, IsAdminAPIEnabledGuard)
-  @ApiOperation({ summary: "Create a private link for a team event type" })
+    @UseGuards(ApiAuthGuard, IsOrgGuard, IsNotPlatformOrgGuard, RolesGuard, IsTeamInOrg, PlatformPlanGuard, IsAdminAPIEnabledGuard)
```
or rather since `IsOrgGuard` attaches org to request then `IsNotPlatformOrgGuard` just like is not managed user could just check if request.organization is platform or not instead of fetching it again

supalarry commented on apps/api/v2/src/ee/event-types-private-links/services/organization-context.service.ts:
```diff
@@ -0,0 +1,36 @@
+import { TeamsEventTypesRepository } from "@/modules/teams/event-types/teams-event-types.repository";
+import { Injectable, NotFoundException } from "@nestjs/common";
+
+export interface OrganizationEventTypeContext {
```
I don't see this interface being used anywhere - do we need it or can we delete it all together since typescript function `getOrganizationEventTypeContext` return type is inferred automatically anyway. Very least i would remove export of this interface not to think whether or not its used somewhere else or not.

supalarry reviewed (changes requested):
Have some comments about possibly redundant org fetching

hariombalhara committed:
fix module deps

ThyMinimalDev committed:
fix: module imports

ThyMinimalDev committed:
fix: remove orgContextService

ThyMinimalDev committed:
fix: add env example

ThyMinimalDev committed:
fix e2e test

ThyMinimalDev committed:
Merge branch 'main' into devin/org-team-event-types-private-links-1762193767

ThyMinimalDev committed:
fixup! fix e2e test

ThyMinimalDev committed:
fix e2e test

ThyMinimalDev committed:
Merge branch 'main' into devin/org-team-event-types-private-links-1762193767

hariombalhara reviewed (changes requested):
v2 Tests are failing

ThyMinimalDev committed:
Merge branch 'main' into devin/org-team-event-types-private-links-1762193767

devin-ai-integration[bot] committed:
Merge branch 'main' into devin/org-team-event-types-private-links-1762193767

Co-Authored-By: unknown <>

devin-ai-integration[bot] committed:
Merge branch 'main' into devin/org-team-event-types-private-links-1762193767

Co-Authored-By: unknown <>
